Gutenboarding: update site creation flow

razvanpapadopol opened the pull request:
#### Design and functional updates:
* Add Domains step and render Domains Modal as a Step.
* Render Plans Modal as a Step.
* If the user enters a site title on Intent Capture then show Domain step next, before Design Selection.
* If the user skips entering a site title, then show Domain step after Style Preview (before Plans).
* Remove Continue button from Plans Page and submit on Select.
* Remove pre-selected plan from Plans page.
* Hide PlansButton and DomainPickerButton from Intent Capture, Domains and Plans steps.
* Align intent capture CTA to right on mobile
* Change copy for paid domains price label.
* Add max-width to Gutenboarding except design selection and create-site.

#### Technical updates:
* Add useStepNavigation hook and use it in Gutenboarding steps.
* Add BackButton, NextButton and SkipButton to ActionButtons.
* Use only ActionButtons in Gutenboarding for consistency.
* Trigger navigation to previous and next steps in useStepNavigation hook.
* Add useOnLogin and useSignup hooks and extract logic from header.
* Update abtest slug
* Fix Gutenboarding e2e tests

####  Analytics events update
  -  `calypso_newsite_modal_open` with prop `{ name: 'PlansGrid' }` becomes `calypso_newsite_step_enter` with prop `{ step: 'PlansModal' }`.
  -  `calypso_newsite_modal_close` with prop `{ name: 'PlansGrid' }` becomes `calypso_newsite_step_leave` with prop `{ step: 'PlansModal' }`.

#### Follow up:
- Register new analytics events and changes: https://github.com/Automattic/wp-calypso/issues/42250

#### Testing instructions

* Go to [/new](https://calypso.live/new?branch=update/gutenboarding-site-creation-flow)
* Test whole Gutenboarding flow and check for the changes mentioned above.
* Check analytics event to fire as before with the changes mentioned above.
* Check different routes including i18n and plan slugs (eg: `/new/design/business/es`).

Fixes https://github.com/Automattic/wp-calypso/issues/43134 and https://github.com/Automattic/wp-calypso/issues/43133
Fixes #43205

Commit:
- Link PlansButton to /plans route with fromPlansButton flag on location state.
- Add onPlanSelect to PlansGrid.
- Remove Continue button from Plans Page and submit on Select.
- Update analytics TracksPlanSelectEventProperties with from_plans_button prop.

Commit:
Remove PlansModal from Gutenboarding.

Commit:
Hide PlansButton from Intent Capture and Plans steps.

Commit:
Remove pre-selected plan from Plans page.

Commit:
Update gutenboarding abtest slug.

Commit:
- Add PlansModel as a Step.
- Use separate route /plans-modal to render the same PlansStep component.
- Record PlansModal as a step name when calling analytics useTrackStep.

Commit:
Don't trigger onClose callback from DomainPicker when pressing Confirm.

Commit:
Add useStepNavigation hook and use it in Gutenboarding steps.

Commit:
- Add DomainsStep and render it on /domains route.
- Remove DomainModal and DomainPopover and display DomainsStep on /domain-modal route.
- Update header to stop rendering DomainPickerButton and PlansButton on certain routes.

simison commented on client/lib/abtest/active-tests.js:
```diff
@@ -127,7 +127,7 @@ export default {
 	},
 	newSiteGutenbergOnboarding: {
-		datestamp: '20200605',
+		datestamp: '20200610',
```
Just leaving note to update this to correct date before merging.

simison commented:
Let's make sure to apply some max-width here:

<img width="2356" alt="image" src="https://user-images.githubusercontent.com/87168/84364628-00ece700-abd9-11ea-9370-59f44d369827.png">

@dubielzyk @olaolusoga what would be good? Both to the buttons and the list itself?

dubielzyk commented:
Yep, a max-width would be good. Have it apply on the whole thing at like 1400px or whatever you see fit? I can't test on a larger monitor

simison commented:
@razvanpapadopol let's also apply the list-borders before merging:

<img width="1348" alt="image" src="https://user-images.githubusercontent.com/87168/84365823-aeacc580-abda-11ea-8129-6280fd5bc42c.png">

Note hover-background on `.net` row.

razvanpapadopol commented:
@dubielzyk you can test using _Developer Tools => device toolbar_ in Chrome.

This is how it looks in production.
<img width="1315" alt="Screenshot 2020-06-11 at 11 55 14" src="https://user-images.githubusercontent.com/14192054/84365668-77d6af80-abda-11ea-86b1-d039ea6b4596.png">

Where exactly should we apply max-width?

dubielzyk commented:
The whole content (everything except the top bar) should have a max-width

alshakero commented on client/landing/gutenboarding/components/header/index.tsx:
```diff
@@ -150,12 +149,13 @@ const Header: React.FunctionComponent = () => {
-		} );
-		setDomain( suggestion );
-	};
+	const showFullHeader = ! [
+		'IntentGathering',
+		'Domains',
+		'DomainsModal',
+		'Plans',
+		'PlansModal',
+	].includes( currentStep );
```
I don't fully understand this. I think it's more intuitive to use positive conditions (as opposed to negative). 

```suggestion
	const showFullHeader = [
		'DesignSelection',
		'Style',
		'Signup',
		'Login',
		'CreateSite',
	].includes( currentStep );
```

razvanpapadopol replied:
I wanted to maintain a list of exceptions 🤔 
Maybe it would be better to name it `const hideFullHeader`?

alshakero replied:
Yup that also works.

alshakero commented on client/landing/gutenboarding/hooks/use-step-navigation.ts:
```diff
@@ -0,0 +1,37 @@
+	const currentStep = useCurrentStep();
+	const siteTitle = useSelect( ( select ) => select( ONBOARD_STORE ).getSelectedSiteTitle() );
+
+	const steps = siteTitle
+		? [ Step.IntentGathering, Step.Domains, Step.DesignSelection, Step.Style, Step.Plans ]
+		: [ Step.IntentGathering, Step.DesignSelection, Step.Style, Step.Domains, Step.Plans ];
```
Could you please add a comment explaining the business logic behind this? Very well implemented btw.

alshakero commented on client/landing/gutenboarding/onboarding-block/acquire-intent/index.tsx:
```diff
@@ -51,8 +50,8 @@ const AcquireIntent: React.FunctionComponent = () => {
 
-	// translators: Button label for advancing to Design Picker step in onboarding
-	const nextLabel = __( 'Choose design' );
+	// translators: Button label for advancing to second step in onboarding
```
```suggestion
	// translators: Button label for advancing to the second step in onboarding
```

razvanpapadopol replied:
made it generic now as part of [ActionButtons](https://github.com/Automattic/wp-calypso/pull/43136/files#diff-6ae3fac2fb17d29ef4858d096d249a13R58)

alshakero commented on client/landing/gutenboarding/onboarding-block/domains/index.tsx:
```diff
@@ -0,0 +1,91 @@
+		selected_domain: domain?.domain_name,
+	} ) );
+
+	const handleBack = () => ( isModal ? history.goBack() : history.push( previousStepPath ) );
```
Brilliant.

alshakero reviewed (approved):
This is so well-implemented. The problem solving is great. Great hook implementation. Left a couple minor notes.

Commit:
- Add BackButton, NextButton and SkipButton to ActionButtons.
- Use only ActionButtons in Gutenboarding for consistency.
- Remove BottomBarMobile.

Commit:
- Update Domain page/modal header and action buttons.
- Add header prop and cleanup DomainPicker.

Commit:
Add comment and cleanup header.

Commit:
Fix mobile styles.

andrewserong commented:
This is some epic work @razvanpapadopol, really nice refactors too, and overall it's looking great! The analytics events for domains and plans are looking good to me, confirmed that the `calypso_newsite_step_enter` and leave events appear to be firing correctly for `Domains`, `DomainsModal`, `Plans`, and `PlansModal`.

I noticed a couple of other things while testing:

* The action buttons appear to use a slightly heavier font weight now, which looks intentional. The `Create account` button in the signup modal will need to be updated to use this new weight, too.
* The `Skip for now` button appears to set the outline to transparent when it's active. This makes it difficult to see that the button is active when tabbing on a keyboard:

![skip-button-tabbing](https://user-images.githubusercontent.com/14988353/84458631-d66a5f00-aca8-11ea-8825-21fae82f99cf.gif)

* When I select a plan using the Plans modal, afterwards I'm getting an intermittent rendering issue where a bit of the blue outline appears to be poking through (I'm using Google Chrome on Mac), but I'm not quite sure what's going on there:

![image](https://user-images.githubusercontent.com/14988353/84458749-16314680-aca9-11ea-8bd5-357fb91eea5f.png)

* UX-wise, if I pick a plan from the Plans modal, when I get to the Plans page, it feels confusing that I need to select a plan again. When we were showing which plan was selected on the Plans page, and the user confirms by clicking Continue, this felt like the Plans page was "confirm what you selected", but without that, it feels like an extra step to have to select the plan again. Not sure how big an issue this is, but my personal preference would be to hide showing the plans step again if someone has selected a plan via the Plans modal.

Commit:
Align intent capture CTA to right on mobile

Commit:
Fix DomainModal skip behaviour.

Commit:
Hide Domains and Plans header instead of unmount.

andrewserong commented:
Thanks @razvanpapadopol https://github.com/Automattic/wp-calypso/pull/43136/commits/2f74d94c3465661746e4df3153d8695b868db435 appears to have fixed the rendering issue with the plans button at the top right of the screen 👍

Commit:
use default font weight for action buttons

Commit:
Trigger navigation to previous and next steps in useStepNavigation hook.

Commit:
Change copy for paid domains price label.

Commit:
- Add useOnLogin and useSignup hooks and extract logic from header.
- Render SignupForm in Gutenboard component.
- Cleanup domain suggestion logic from header.

Commit:
- Add showSignupDialog to ONBOARD store and use it in useStepNavigation hook.
- Cleanup signup and site creation logic from components

Commit:
re-add grid style to domain picker

Commit:
add max-width to Gutenboarding except design selection and create-site

Commit:
update abtest slug

razvanpapadopol commented:
Thanks @andrewserong for review! 💯 
I've made also the following changes:
- use default font-weight on every button.
- selecting a domain using Domain button on Plans button in header will remove Domains step or Plans step at the end of the flow.

Commit:
Fix Gutenboarding e2e tests

alshakero reviewed (commented):
Great work! Beautiful hook designs and much cleaner structure. I just wish it was split into smaller PRs.

I reviewed the code and tested from start to end.

I found that on mobile, the domain picker has some flexing issues:
![hash-6c1c5742bec06b341eb85582facdb4402f3c453e calypso live_new_domains(iPhone X)](https://user-images.githubusercontent.com/17054134/84521834-e482a480-acd5-11ea-9a05-271b09fe1509.png)

And that IE is broken. 

I have to say I couldn't focus enough to give a good review. I'll give it another one once the IE issue is fixed, if you agree it should be.

Commit:
Adjust size of paid domain price labels on mobile.

Commit:
Fix IE11 visual bug because of unsupported display: grid
